stellargraph/data/explorer.py

```diff
@@ -774,45 +599,42 @@ def run(self, nodes=None, n=1, n_size=None, seed=None):
+        rs = self._check_parameter_values(nodes, n, n_size, seed)
 
         walks = []
-        d = len(n_size)  # depth of search
-
-        if seed:
-            # seed the random number generator
-            rs = random.Random(seed)
-        else:
-            # Restore the random state
-            rs = self._random_state
+        max_hops = len(n_size)  # depth of search
 
         for node in nodes:  # iterate over root nodes
             for _ in range(n):  # do n bounded breadth first walks from each root node
                 q = list()  # the queue of neighbours
                 walk = list()  # the list of nodes in the subgraph of node
                 # extend() needs iterable as parameter; we use list of tuples (node id, depth)
-                q.extend([(node, 0)])
+                q.append((node, 0))
 
                 while len(q) > 0:
                     # remove the top element in the queue
                     # index 0 pop the item from the front of the list
-                    frontier = q.pop(0)
-                    depth = frontier[1] + 1  # the depth of the neighbouring nodes
-                    walk.extend([frontier[0]])  # add to the walk
-
-                    # consider the subgraph up to and including depth d from root node
-                    if depth <= d:
-                        neighbours = self.neighbors(frontier[0])
-                        if len(neighbours) == 0:
-                            break
-                        else:
-                            # sample with replacement
-                            neighbours = [
-                                rs.choice(neighbours) for _ in range(n_size[depth - 1])
-                            ]
+                    cur_node, cur_depth = q.pop(0)
+                    depth = cur_depth + 1  # the depth of the neighbouring nodes
+                    walk.append(cur_node)  # add to the walk
+
+                    # consider the subgraph up to and including max_hops from root node
+                    if depth > max_hops:
+                        continue
+                    neighbours = (
+                        self.neighbors(cur_node) if cur_node is not None else []
+                    )
+                    if len(neighbours) == 0:
+                        # Either node is unconnected or is in directed graph with no out-nodes.
+                        neighbours = [None] * n_size[cur_depth]
+                    else:
+                        # sample with replacement
+                        neighbours = [
+                            rs.choice(neighbours) for _ in range(n_size[cur_depth])
+                        ]
 
-                        # add them to the back of the queue
-                        q.extend([(sampled_node, depth) for sampled_node in neighbours])
+                    # add them to the back of the queue
+                    q.extend([(sampled_node, depth) for sampled_node in neighbours])
```

Comment:
The intermediate list isn't necessary, if you're refactoring this code:

```suggestion
                    q.extend((sampled_node, depth) for sampled_node in neighbours)
```

Reply:
Didn't know that one!